dkg/dkg_test.go

```diff
@@ -405,3 +408,134 @@ func startNewDKG(t *testing.T, parentCtx context.Context, config dkg.Config, dkg
 
 	return cancel
 }
+
+// TestKeymanager tests the keymanager functionality where the dkg command pushes validator keyshares
+// to a keymanager API url.
+func TestKeymanager(t *testing.T) {
+	numNodes := 4
+	lock, p2pKeys, _ := cluster.NewForT(t, 1, 3, numNodes, 0)
+	def := lock.Definition
+	require.NoError(t, def.VerifySignatures())
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Setup
+	dir, err := os.MkdirTemp("", "")
+	require.NoError(t, err)
+
+	allReceived := make(chan struct{}, 1) // Receives struct{} if all `numNodes` keystores are intercepted by the keymanager server
+	srv := httptest.NewServer(newKeymanagerHandler(t, numNodes, allReceived))
+	defer srv.Close()
+
+	// Start relay.
+	relayAddr, errChan := startRelay(ctx, t)
+
+	conf := dkg.Config{
+		DataDir: dir,
+		P2P: p2p.Config{
+			Relays: []string{relayAddr},
+		},
+		KeymanagerAddr: srv.URL,
+		Log:            log.DefaultConfig(),
+		TestDef:        &def,
+	}
+
+	// Run dkg for each node
+	var eg errgroup.Group
+	for i := 0; i < len(def.Operators); i++ {
+		conf := conf
+		conf.DataDir = path.Join(dir, fmt.Sprintf("node%d", i))
+		conf.P2P.TCPAddrs = []string{testutil.AvailableAddr(t).String()}
+
+		require.NoError(t, os.MkdirAll(conf.DataDir, 0o755))
+		err := crypto.SaveECDSA(p2p.KeyPath(conf.DataDir), p2pKeys[i])
+		require.NoError(t, err)
+
+		eg.Go(func() error {
+			err := dkg.Run(ctx, conf)
+			if err != nil {
+				cancel()
+			}
+
+			return err
+		})
+		if i == 0 {
+			// Allow node0 some time to startup, this just mitigates startup races and backoffs but isn't required.
+			time.Sleep(time.Millisecond * 100)
+		}
+	}
+
+	// Wait until complete
+
+	runChan := make(chan error, 1)
+	go func() {
+		runChan <- eg.Wait()
+	}()
+
+	select {
+	case err := <-errChan:
+		// If this returns first, something went wrong with the relay and the test will fail.
+		cancel()
+		testutil.SkipIfBindErr(t, err)
+		require.Fail(t, "bootnode error: %v", err)
+	case err := <-runChan:
+		cancel()
+		testutil.SkipIfBindErr(t, err)
+		require.NoError(t, err)
+	}
+
+	// Wait until all keystores are received by the keymanager server
+	<-allReceived
+	t.Log("All keystores received 🎉")
+}
+
+// noopKeymanagerReq is a mock keystore.keymanagerReq for use in tests.
+type noopKeymanagerReq struct {
+	Keystores []noopKeystore `json:"keystores"`
+	Passwords []string       `json:"passwords"`
+}
+
+// noopKeystore is a mock keystore.keystore for use in tests.
+type noopKeystore struct {
+	Crypto map[string]interface{} `json:"crypto"`
+}
+
+// decrypt returns the secret from the encrypted keystore.
+func decrypt(t *testing.T, store noopKeystore, password string) (*bls_sig.SecretKey, error) {
```

Comment:
this feels like testing the keymanager client, which isn't required, you test that implementation in the keymanager package, this test is for DKG, testing that is submits something is fine.

Reply:
cool, removing this. I'm now only ensuring that the request body received by the server is non-empty